### DISCUSSION TITLE
SD card recording constant droning / clicking noise.

### DISCUSSION THREAD

**Discussion by GuyTraveler** [pschatzmann/arduino-audiokit#17] [Q&A]
I have been working with the AI Thinker audio kit board for the past several weeks. Initially I tried using the ESP-ADF project. I was able to get the playback, keys, and SD card recording examples to work. However, the recorded audio had a constant terrible droning / clicking noise present. I exhausted my configuration / debugging options. I decided to try and leverage this Arduino implementation. My board is identified as V2.2 A149 with 2974 marked on the back of the ESP32 module. I am using #define AUDIOKIT_BOARD 5 for configuration. I have tried different SD card libraries, different settings for the SD card library, different buffer sizes, different encoders, different recording configurations, etc. I have scoured the es8388 driver file and implementation files for possible resolutions. I have varied the board configuration around a number of variables including SPI flash and clock speed. The significant droning / clicking sound still persists. I am unsure about how to proceed.

Do you have any ideas how to resolve this issue?

**Comment by pschatzmann** [comment]
If you have this issue both with ESP-ADF  and my library I would rather think that you have a defective board.
Do you have the issue both with the speaker and the headphone ?

**Reply by iotregenesis** [reply]
hi all,
excuse me for jumping in to your discussion.

I had probably the same issue with the clicking sound while playing from SD Card. 
I found out that all my mp3 files with 44100 sampling rate introduced those clicking sound while being played.
I down-sampled them and I noticed that those clicking noises are gone if the audio file is sampled with as high as 32000 rate.

I used this example project from Phil to test:
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd-audiokit/player-sd-audiokit.ino
with this 
build_flags = 
    -DCORE_DEBUG_LEVEL=2 
    -DAUDIOKIT_BOARD=7 
    -DAI_THINKER_ES8388_VOLUME_HACK=1 

Now, does anyone know if there a way to make those files with 44100 sampling rate be played without clicking noise?

/fahmy

**Reply by iotregenesis** [reply]
ah ,,, never mind guys.
I will try to follow the suggestion [below](https://github.com/pschatzmann/arduino-audiokit/discussions/17#discussioncomment-3506703) 

cheers

**Reply by pschatzmann** [reply]
Did you try out other mp3 decoders ?

**Reply by iotregenesis** [reply]
No. I have not. Currently I am using HELIX.
I will try MAD or MINI later.

edit:
I just tried with libMAD. 
Playing mp3 file with 44100 sample rate and 160 kpbs is good with libhelix but clicking with libmad.
Playing mp3 file with 44100 sample rate and 128 kpbs is good with libhelix and also good with libmad.

**Comment by GuyTraveler** [comment]
I have only tested with headphones. All the example programs and associated pre-recorded audio files playback with no issues.

I am certainly not ruling out that the board my have a defect but I do not have a way to conclude that especially considering the number of weird issues discussed throughout the ESP-ADF repository, your repositories, and general forums. (Droning noise, clicking sound, poor quality, noisy, and other keywords that seem similar to my situation.)

**Comment by pschatzmann** [comment]
I would suggest to test with a simple sine tone to rule out that the issue is with your audio data.
I also suggest to test with a speaker: I did not really use any headphones... 
You could also try to power it from an accu or batteries to rule out that your issue comes from a noisy power supply.

I never had or heard of your issue, so I don't really know...

**Comment by GuyTraveler** [comment]
Thank you for all of your debugging ideas. I will work through them. I was considering it my be related to a noisy power supply. Just for clarity what is an "accu"? Thanks again.

**Comment by pschatzmann** [comment]
Sorry, I meant a capacitor to stabilize the power....

**Reply by GuyTraveler** [reply]
I tested recording data to the SD card by using the SineWaveGenerator (including varying the frequency) and SilenceGenerator. I did not notice any droning or clicking noise. I also took your suggestion and used a variety of means to power the board. (I varied the cable and power source [wall adapters / USB battery banks]) The audio still contained a droning / clicking sound. For reference I am including a snipped image containing the Serial Plotter data when there is no sound in the environment. I am recording from Line2 which based on your blog posts is a composite of the LineIn and the two on board microphones. 
![arduinoplotter_output](https://user-images.githubusercontent.com/5072937/165688759-d8b153f8-fdd5-4b82-b680-615f81dad372.png)

**Comment by pschatzmann** [comment]
I am starting to get confused. 
- Did you test the audio output (e.g. with a sine wave or playing a file from the sd card or URL ?
- If you want to test the recording you would copy the input from the microphone to a file and then check the quality of that file

Is my assumption correct that your issue is when you copy the audio from the line2 to the audiokit output ?  

I think your print screen is normal and barely audible if you consider that the max range is 32768,,,

**Reply by GuyTraveler** [reply]
I wrote the audio output from the SineWaveGenerator to the SD card. I then reviewed that recording and did not notice any droning / clicking noises. I repeated that process for the SilenceGenerator and the SineWaveGenerator again varying the frequency and received the same results. (A recorded audio file without any droning / clicking noises) 

Yes. The issue occurs when I copy the audio from Line2 to the SD card. 

I would also believe that the small variance / background noise displayed by the Serial Plotter would not be problematic. Unfortunately the noise is so pronounced that it makes the recorded audio unusable. I listen to the recordings on my laptop from the SD card. I have reviewed practically every forum post or GitHub issue / discussion page. It does seem like other users are experiencing similar issues: https://www.mikrocontroller.net/topic/474383 For clarity I do not believe the issue is with your library but some sort of board specific configuration. 

Is there some sort of filter or effect I could apply to remove the noise? I reworked the VolumeStream class and created my own ThresholdStream (threshold / dropout) to remove noise. Even though the plotter only showed around a maximum value of 60 for noise it required a value of 400 to remove the droning / clicking noise. Even then there were still some points of noise that went beyond 400. As well the noise would still be present in the audio above the threshold like my speech. 

I am not familiar with audio processing algorithms but my current effort will be put into processing the audio. Ideally I would like to resolve the noise issue through a configuration setting versus processing.

I have attached a representation of an audio recording opened in Audacity.
![audacity_amplified](https://user-images.githubusercontent.com/5072937/165989503-5fdecc21-9eb5-44dd-9b9c-1f492c750d9c.png)

**Comment by pschatzmann** [comment]
Hmm, If you want to use the microphone, I suggest that you use line1. 
You can also try other microphone gains settings by calling 
esp_err_t es8388_set_mic_gain(es_mic_gain_t gain);

**Reply by GuyTraveler** [reply]
Thank you. I will try adjusting the mic gain value. That is certainly one of the configuration components I initially varied. I also think I tinkered with the underlying code since I noticed a change in the code in the ESP-ADF repository: gain_n = (gain_n << 4) + gain_n;
Although now I have stumbled across a separate repository that performs a bitwise operation instead of a plus operation: (gain_n << 4) | gain_n

I will proceed with a two prong approach: 

1. Try and improve the audio quality with processing algorithms.
2. With fresh eyes look over the es8388.c file and its usage to determine if the audio can be improved through configuration.

Thanks again for your help. I will let you know if I have success.

**Comment by pschatzmann** [comment]
This might be of interrest: https://datasheet.lcsc.com/lcsc/1912111437_Everest-semi-Everest-Semiconductor-ES8388_C365736.pdf

**Reply by TangerineDreamer** [reply]
Seems i had good playing with with 44.100khz and 160kbps mp3 files on the SD card
but bad glitching playing with 360 kbps,i'm sure that the codec limitations.

**Reply by GuyTraveler** [reply]
I appreciate all your help. I wanted to get a reply back to you. I did use that documentation to modify codec configuration to try and improve the audio quality. Unfortunately I could not get an improvement. I investigated using a noise removal algorithm and abandoned the effort because of time constraints. I ordered another AiThinker board assuming it might be a defect with my existing board. The board I received had the same issue. Finally I decided to purchase the LyraT v4.3 board by Espressif. I have had no issue with audio quality when recording with the on board microphones. The audio quality is excellent. If I am going to guess I think there is some interference on the aux in on the AiThinker board. Since the aux in and on board microphones input is mixed and then amplified it is more pronounced. From here on out it seems to be mostly application development tasks. 

Again thanks for your timely responses and openness to helping.

**Reply by TangerineDreamer** [reply]
Hey GUY,why don't you try to convert an audio file to mp3,128kbs bitrate and another to mp3,320kbs 
and listen to the difference between them.I was strugging like you to find out what's going on,and
now i've no glitches or cliks and crocks and no pop corns too :)
I think 160kbs is a  good limit,128kbs is exellent but audio frequency  limited. 
If you have exellent audio with microphone,it could be that the recording bitrate is pre-limited
to something under 160bks,could be 128kbs
